Passing static constructors of Eigen-types to pinocchio methods

Hi all,
this is not an urgent issue, but for the sake of documentation, let me note it here:

For newer Eigen versions (tested with 3.3.5 and 3.3.3), there is a compilation error when passing statically constructed arguments to pinocchio methods. 

For example, if you try to compile
`pinocchio::forwardKinematics(model_, data_, JointPosition_t::Zero());`

you will obtain the compiler-error
`
/usr/local/include/pinocchio/multibody/joint/joint-free-flyer.hpp:244:95: error: ‘Options’ is not a member of ‘Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 7, 1, 0, 7, 1> >’
       typedef typename Eigen::Quaternion<typename ConfigVector::Scalar,ConfigVector::Options> Quaternion;
                                                                                               ^
/usr/local/include/pinocchio/multibody/joint/joint-free-flyer.hpp:245:44: error: ‘Options’ is not a member of ‘Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 7, 1, 0, 7, 1> >’
       typedef Eigen::Map<const Quaternion> ConstQuaternionMap;
`

If you instead use the argument
`const JointPosition_t zeroPos = JointPosition_t::Zero();
pinocchio::forwardKinematics(model_, data_, zeroPos);`
compilation works fine.

--- a/src/multibody/joint/joint-free-flyer.hpp
+++ b/src/multibody/joint/joint-free-flyer.hpp
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2015-2018 CNRS
+// Copyright (c) 2015-2019 CNRS INRIA
 // Copyright (c) 2015-2016 Wandercraft, 86 rue de Paris 91400 Orsay, France.
 //
 
@@ -239,19 +239,41 @@ namespace pinocchio
       M.translation(q_joint.template head<3>());
     }
     
+    template<typename Vector3Derived, typename QuaternionDerived>
+    EIGEN_DONT_INLINE
+    void calc(JointDataDerived & data,
+              const typename Eigen::MatrixBase<Vector3Derived> & trans,
+              const typename Eigen::QuaternionBase<QuaternionDerived> & quat) const
+    {
+      data.M.translation(trans);
+      data.M.rotation(quat.matrix());
+    }
+    
     template<typename ConfigVector>
     EIGEN_DONT_INLINE
     void calc(JointDataDerived & data,
-              const typename Eigen::MatrixBase<ConfigVector> & qs) const
+              const typename Eigen::PlainObjectBase<ConfigVector> & qs) const
     {
       typedef typename Eigen::Quaternion<typename ConfigVector::Scalar,ConfigVector::Options> Quaternion;
       typedef Eigen::Map<const Quaternion> ConstQuaternionMap;
       
       typename ConfigVector::template ConstFixedSegmentReturnType<NQ>::Type q = qs.template segment<NQ>(idx_q());
-      data.M.translation(q.template head<3>());
-      
       ConstQuaternionMap quat(q.template tail<4>().data());
-      data.M.rotation(quat.matrix());
+      
+      calc(data,q.template head<3>(),quat);
+    }
+    
+    template<typename ConfigVector>
+    EIGEN_DONT_INLINE
+    void calc(JointDataDerived & data,
+              const typename Eigen::MatrixBase<ConfigVector> & qs) const
+    {
+      typedef typename Eigen::Quaternion<Scalar,Options> Quaternion;
+      
+      typename ConfigVector::template ConstFixedSegmentReturnType<NQ>::Type q = qs.template segment<NQ>(idx_q());
+      const Quaternion quat(q.template tail<4>());
+      
+      calc(data,q.template head<3>(),quat);
     }
     
     template<typename ConfigVector, typename TangentVector>

--- a/src/multibody/joint/joint-spherical.hpp
+++ b/src/multibody/joint/joint-spherical.hpp
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2015-2018 CNRS
+// Copyright (c) 2015-2019 CNRS INRIA
 // Copyright (c) 2015-2016 Wandercraft, 86 rue de Paris 91400 Orsay, France.
 //
 
@@ -425,18 +425,35 @@ namespace pinocchio
       M.rotation(quat.matrix());
       M.translation().setZero();
     }
-
-    template<typename ConfigVector>
+    
+    template<typename QuaternionDerived>
     void calc(JointDataDerived & data,
-              const typename Eigen::MatrixBase<ConfigVector> & qs) const
+              const typename Eigen::QuaternionBase<QuaternionDerived> & quat) const
+    {
+      data.M.rotation(quat.matrix());
+    }
+    
+    template<typename ConfigVector>
+    EIGEN_DONT_INLINE
+    void calc(JointDataDerived & data,
+              const typename Eigen::PlainObjectBase<ConfigVector> & qs) const
     {
       typedef typename Eigen::Quaternion<typename ConfigVector::Scalar,ConfigVector::Options> Quaternion;
       typedef Eigen::Map<const Quaternion> ConstQuaternionMap;
-      
-      typename ConfigVector::template ConstFixedSegmentReturnType<NQ>::Type & q = qs.template segment<NQ>(idx_q());
-      
-      ConstQuaternionMap quat(q.data());
-      data.M.rotation(quat.matrix());
+
+      ConstQuaternionMap quat(qs.template segment<NQ>(idx_q()).data());
+      calc(data,quat);
+    }
+
+    template<typename ConfigVector>
+    EIGEN_DONT_INLINE
+    void calc(JointDataDerived & data,
+              const typename Eigen::MatrixBase<ConfigVector> & qs) const
+    {
+      typedef typename Eigen::Quaternion<Scalar,Options> Quaternion;
+
+      const Quaternion quat(qs.template segment<NQ>(idx_q()));
+      calc(data,quat);
     }
 
     template<typename ConfigVector, typename TangentVector>

--- a/unittest/kinematics.cpp
+++ b/unittest/kinematics.cpp
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2018 CNRS
+// Copyright (c) 2018-2019 CNRS INRIA
 //
 
 #include "pinocchio/multibody/model.hpp"
@@ -15,6 +15,23 @@
 #include <boost/utility/binary.hpp>
 
 BOOST_AUTO_TEST_SUITE(BOOST_TEST_MODULE)
+
+BOOST_AUTO_TEST_CASE(test_kinematics_constant_vector_input)
+{
+  using namespace Eigen;
+  using namespace pinocchio;
+  
+  Model model;
+  buildModels::humanoidRandom(model);
+  
+  Data data(model);
+  
+  model.lowerPositionLimit.head<3>().fill(-1.);
+  model.upperPositionLimit.head<3>().fill(1.);
+  VectorXd q = randomConfiguration(model);
+  
+  forwardKinematics(model,data,Model::ConfigVectorType::Ones(model.nq));
+}
 
 BOOST_AUTO_TEST_CASE(test_kinematics_zero)
 {

